src/App.tsx

```diff
@@ -43,25 +43,135 @@ const CustomListView = ({ program, dispatch }) => {
+  deckNames?: string[];
+  subject: Subject;
+};
+
+type State = {
+  view: View;
+  program: PROGRAM;
+  cardEditor?: CodeMirror.Editor;
+  subjectToList: Subject;
+  deckToViewDetail?: string;
+  complexCommandParams?: ComplexCommandParams;
+};
+
+export enum ActionType {
+  SetCardEditor = "set card editor",
+  CardEditorParseSuccess = "card editor parse success",
+  StartSession = "start session",
+  List = "list",
+  ViewDeckDetail = "view deck detail",
+  ShowStats = "show stats",
+  LoadDecks = "load decks",
+  CommandNotFound = "command not found",
+}
+
+export type Action =
+  | {
+      type: ActionType.SetCardEditor; // Enables "Load decks" to work
+      cardEditor: CodeMirror.Editor;
+    }
+  | {
+      type: ActionType.CardEditorParseSuccess;
+      program: PROGRAM;
+    }
+  | {
+      type: ActionType.StartSession;
+      limit?: number;
+      filter?: Filter;
+      isLimitAppliedToCards?: boolean;
+      deckNames?: string[];
+      subject: Subject;
+    }
+  | {
+      type: ActionType.List;
+      listOption: "decks" | "tags";
+      // tags not implemented
+    }
+  | {
+      type: ActionType.ViewDeckDetail;
+      deckName: string;
+    }
+  | {
+      type: ActionType.ShowStats;
+      limit: number;
+      filter?: Filter;
+      isLimitAppliedToCards?: boolean;
+      deckNames?: string[];
+      subject: Subject;
+    }
+  | {
+      type: ActionType.LoadDecks;
+      createDSLValue: string;
+    }
+  | {
+      type: ActionType.CommandNotFound;
+    }; // Add actions here
+
+const reducer = (state: State, action: Action): State => {
+  debug("Dispatched to App: ", action);
   switch (action.type) {
+    // Add action handlers here
+    case "set card editor": {
+      return {
+        ...state,
+        cardEditor: action.cardEditor,
+      };
+    }
     case "card editor parse success": {
       return {
         ...state,
-        program: { ...action.program },
+        program: action.program,
         view: View.DECK,
       };
     }
     case "start session": {
       return {
         ...state,
         view: View.SESSION,
-        from: {
+        complexCommandParams: {
           limit: action.limit,
           filter: action.filter,
-          selectedCards: action.selectedCards, // determines whether limit applies to cards or SUBJECT (decks/sessions)
-          deckNames: action.deckNames, // if deckNames is null then it is from past sessions
+          isLimitAppliedToCards: action.isLimitAppliedToCards,
+          deckNames: action.deckNames,
+          subject: action.subject,
+        },
+      };
+    }
+    case "show stats": {
+      return {
+        ...state,
+        view: View.STATS,
+        complexCommandParams: {
+          limit: action.limit,
+          filter: action.filter,
+          isLimitAppliedToCards: action.isLimitAppliedToCards,
+          deckNames: action.deckNames,
+          subject: action.subject, // if deckNames is null then it is from past sessions
```

Comment:
@ZhiHanL  I changed this part significantly:
`selectedCards` was confusing so I renamed it to `isLimitAppliedToCards?: boolean`
I also moved all the selection info for showing stats into `complexCommandParams`
`subject` has enum values (from App.tsx above)
`filter` also has enum values (from query.ts)